sql/sql_acl.cc

```diff
@@ -8651,110 +8652,134 @@ bool mysql_grant_role(THD *thd, List <LEX_USER> &list, bool revoke)
+  List_iterator<LEX_USER> it(list);
+  bool result= false;
+  DBUG_ENTER("mysql_grant_global");
 
-  if (lower_case_table_names && db)
+  /* go through users in user_list */
+  for (LEX_USER *user= it++; user != NULL; user= it++)
   {
-    char *end= strnmov(tmp_db,db, sizeof(tmp_db));
-    if (end >= tmp_db + sizeof(tmp_db))
+    /* Resolve if it the proxy is "CURRENT_USER" or "CURRENT_ROLE" or
+       a username / rolename without a hostname specified. */
+    LEX_USER *resolved_user= get_current_user(thd, user, false);
+    if (!resolved_user)
     {
-      my_error(ER_WRONG_DB_NAME ,MYF(0), db);
-      DBUG_RETURN(TRUE);
+      result= true;
+      continue;
     }
-    my_casedn_str(files_charset_info, tmp_db);
-    db=tmp_db;
+    /* TODO(cvicentiu) This might actually not be necessary for grant_proxy.
+       Carry over auth information. */
+    resolved_user->auth= user->auth;
+    if (check_if_auth_can_be_changed(resolved_user, thd) ||
+        replace_user_table(thd, tables.user_table(),
+                           resolved_user,
+                           priv_spec,
+                           create_new_users,
+                           !is_public(resolved_user) && no_auto_create_users))
+      result= true;
+
+    if (resolved_user->is_role())
+      propagate_role_grants(find_acl_role(resolved_user->user.str, true),
+                            PRIVS_TO_MERGE::GLOBAL);
   }
+  DBUG_RETURN(result);
+}
 
-  if (is_proxy)
+static bool mysql_grant_db(THD *thd,
+                           Grant_tables &tables,
+                           List<LEX_USER> &list,
+                           const char *db,
+                           const Priv_spec &priv_spec,
+                           bool create_new_users,
+                           bool no_auto_create_users)
+{
+  List_iterator<LEX_USER> it(list);
+  bool result= false;
+  DBUG_ENTER("mysql_grant_db");
+  mysql_mutex_assert_owner(&acl_cache->lock);
+  DBUG_ASSERT((priv_spec.access & DB_ACLS) == priv_spec.access);  // Only DB_ACLs are accepted.
+  /* go through users in user_list */
+  for (LEX_USER *user= it++; user != NULL; user= it++)
   {
-    DBUG_ASSERT(!db);
-    proxied_user= str_list++;
-  }
+    /* Resolve if it the proxy is "CURRENT_USER" or "CURRENT_ROLE" or
+       a username / rolename without a hostname specified. */
+    LEX_USER *resolved_user= get_current_user(thd, user, false);
+    if (!resolved_user)
+    {
+      result= true;
+      continue;
+    }
+    /* TODO(cvicentiu) This might actually not be necessary for grant_proxy.
+       Carry over auth information. */
+    resolved_user->auth= user->auth;
+    if (check_if_auth_can_be_changed(resolved_user, thd) ||
+        replace_user_table(thd, tables.user_table(),
+                           resolved_user,
+                           {NO_ACL, priv_spec.revoke},
+                           create_new_users,
+                           !is_public(resolved_user) && no_auto_create_users) ||
```

Comment:
why not to move this (common) part to a separate function? find a user, copy auth, create a user if needed — all your new mysql_grant_* functions do that

Reply:
also the loop over users, propagation of role grants

Reply:
as a summary, there's quite a lot of duplicated code now. I feel that it can be isolated even further